hathor/cli/db_export.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import struct
+from argparse import ArgumentParser, FileType, Namespace
+from typing import Iterator, Optional
+
+from hathor.cli.run_node import RunNode
+
+MAGIC_HEADER = b'HathDB'
+
+
+class DbExport(RunNode):
+    def start_manager(self, args: Namespace) -> None:
+        pass
+
+    def register_signal_handlers(self, args: Namespace) -> None:
+        pass
+
+    def register_resources(self, args: Namespace) -> None:
+        pass
+
+    def create_parser(self) -> ArgumentParser:
+        from hathor.conf import HathorSettings
+        settings = HathorSettings()
+
+        def max_height(arg: str) -> Optional[int]:
+            if arg.lower() == 'checkpoint':
+                if not settings.CHECKPOINTS:
+                    raise ValueError('There are no checkpoints to use')
+                return settings.CHECKPOINTS[-1].height
+            elif arg:
+                return int(arg)
+            else:
+                return None
+
+        parser = super().create_parser()
+        parser.add_argument('--export-file', type=FileType('wb', 0), required=True,
+                            help='Save the export to this file')
+        parser.add_argument('--export-iterator', choices=['fast2', 'fast', 'slow'], default='fast2',
+                            help='Which method of iterating to use, don\'t change unless you know what it does')
+        parser.add_argument('--export-max-height', type=max_height,
+                            help='Stop exporting after a certain height, use "checkpoint" stop at the last one')
```

Comment:
block height? What happens with the mempool?

Reply:
Got it. Maybe we should clarify it on the help message.

Reply:
I tried to clarify, but I also realized it's not so cut and dry. I don't really know what to clarify, actually. The original mempool may or may not be partially exported when using a max-height, and the resulting export file, may or may not contain transactions that will be in the mempool. It mostly depends on the timestamps and on the iterator used (I suspect that when using the dfs iterator and a max-height, the mempool will always be empty, but I'm also not 100% sure). Let me know if you have any suggestions of what to add to the help explanation, because I think there are too many details to explain without turning the help into a full guide.

Most importantly, this command is experimental, so I'm not too worried if the help is unclear or incomplete.

Reply:
Hum, your comments about the mempool make sense. I feel we can use something like: "Make no assumption about the mempool. It may be partially exported or not, depending on the timestamps and the traversal algorithm.".